### ISSUE TITLE
Overriding a RelatedFactoryList at build time doesn't work

### ISSUE THREAD

**Issue by TMiguelT** [FactoryBoy/factory_boy#636]
#### Description
If you call a factory function, and try to pass in a list for a field that is defined as a `RelatedFactoryList`, this list will be empty once the instance is built.

#### To Reproduce
```python
from factory import Factory, RelatedFactoryList, Faker
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, Unicode, create_engine, ForeignKey
from factory.alchemy import SQLAlchemyModelFactory
from sqlalchemy.orm import scoped_session, sessionmaker, relationship

Base = declarative_base()
engine = create_engine('sqlite://')
session = scoped_session(sessionmaker(bind=engine))

class Parent(Base):
    __tablename__ = 'parent'
    id = Column(Integer, primary_key=True)
    children = relationship("Child", back_populates="parent")

class Child(Base):
    __tablename__ = 'child'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey('parent.id'))
    name = Column(Unicode())
    parent = relationship("Parent", back_populates="children")

class ChildFactory(SQLAlchemyModelFactory):
    class Meta:
        model = Child
        sqlalchemy_session = session
    name = Faker('word')

class ParentFactory(SQLAlchemyModelFactory):
    class Meta:
        model = Parent
        sqlalchemy_session = session
    children = RelatedFactoryList(ChildFactory, 'parent', size=1,)

# This works because we let the RelatedFactoryList auto generate children for us
child = ChildFactory()
auto_parent = ParentFactory()
assert len(auto_parent.children) > 0

# This, however, doesn't work
auto_parent = ParentFactory(children = [ChildFactory()])
assert len(auto_parent.children) > 0
```

##### The issue
As far as I understand, the reason this happens is because declarations such as `RelatedFactoryList` are considered a post declaration, and thus run after the construction of the object. However, these post declarations are never actually attached to the instance, only the pre declarations are. I'm not entirely sure why this still works if you let the children be auto-generated, however (ie my `auto_parent = ParentFactory()` example).

The most relevant section of the code seems to be: https://github.com/FactoryBoy/factory_boy/blob/2a2e7b3de5566a8fcd99258bfc9f4762eca2d051/factory/builder.py#L172-L176, where there's a specific code path for passing in values to a post_declaration, but this doesn't seem to be working properly.

**Comment by rbarrois** [comment]
This is a complex topic; things appear to work because ORMs perform some magic under the hood.

With most ORMs, the following holds true (without a factory, just simple objects):
```python
>>> father = Parent(name='anakin')
>>> set(father.children)
set()

>>> children = [Child(name='luke', parent_id=father.id), Child(name='leia', parent_id=father.id)]
>>> set(father.children)
set(<Child: leia>, <Child: luke>)
```

When you're reading `father.children`, the ORM code will fetch all lines whose `parent_id` matches the current object's `id`, and return them.

With most databases, if we enforce that each child object should point to a parent object, it is impossible to write a `Child(name='luke', parent_id=father.id)` to the database before the `father` object has been written — foreign key constraints would fail.
=> In your example, the `ChildFactory()` part of `ParentFactory(children=[...])` would fail in that case, since it's executed *before* the `ParentFactory` has been able to create an object in the database.

This combination of features explains why the `RelatedFactoryList` (and `RelatedFactory`) are both implemented as post-generation: in most cases, the "parent" object must be written **before** its children can exist.

**Comment by rbarrois** [comment]
For the above reasons, I think this is more a documentation problem than a bug or lacking feature: we should provide a good example for handling this typical `Parent  1..n  Child` use case.

**Comment by TMiguelT** [comment]
Hmm, I'm not sure I understand this correctly, but I agree with you that the parent needs to be created before the child.

So, when I run `auto_parent = ParentFactory(children = [ChildFactory()])`, I expect the following to happen:
* A Parent is constructed
* The normal fields are populated according to the Factory
* Because the children field is specified as an argument, no children are constructed
* The already existing children are then attached to the parent

This works at the ORM level:
```python
parent = Parent(children = [Child(**child_args)])
```
Why does it not work here?

**Comment by rbarrois** [comment]
Does it actually work when you save to the database? How does the child's `parent_id` get set to the parent's `id` field?
I'm not that familiar with SQLAlchemy internals; but I can tell it wouldn't work with Django :)

**Comment by TMiguelT** [comment]
Yes, it's a very normal way to use SQLAlchemy: https://docs.sqlalchemy.org/en/13/orm/tutorial.html#working-with-related-objects
